Source/NSGridView.m

```diff
@@ -0,0 +1,957 @@
+  self = [super initWithFrame: frameRect];
+
+  if (self != nil)
+    {
+      _rows = [[NSMutableArray alloc] initWithCapacity: 10];
+      _columns = [[NSMutableArray alloc] initWithCapacity: 10];
+      _cells = [[NSMutableArray alloc] initWithCapacity: 10];
+    }
+
+  [self _refreshCells];
+  
+  return self;
+}
+
+- (instancetype) initWithViews: (NSArray *)rows
+{
+  self = [self initWithFrame: NSZeroRect];
+  
+  if (self != nil)
+    {
+      FOR_IN(NSMutableArray*, row, rows)
+        {
+        }
+      END_FOR_IN(rows);
+    }
+
+  [self _refreshCells];
+  
+  return self;
+}
+
++ (instancetype) gridViewWithNumberOfColumns: (NSInteger)columnCount rows: (NSInteger)rowCount
+{
+  NSUInteger r = 0;
+  NSUInteger c = 0;
+  NSMutableArray *rows = [[NSMutableArray alloc] initWithCapacity: rowCount]; 
+
+  for (r = 0; r < rowCount; r++)
+    {
+      NSMutableArray *col = [NSMutableArray arrayWithCapacity: columnCount];
+      for (c = 0; c < columnCount; c++)
+        {
+          NSGridCell *gc = [[NSGridCell alloc] init];
+          [col addObject: gc];
+          RELEASE(gc);
+        }
+      [rows addObject: col];
+    }
+  
+  return AUTORELEASE([self gridViewWithViews: rows]);
+}
+
++ (instancetype) gridViewWithViews: (NSArray *)rows
+{
+  return [[self alloc] initWithViews: rows];
+}
+
+- (NSInteger) numberOfRows
+{
+  return [_rows count];
+}
+
+- (NSInteger) numberOfColumns
+{
+  return [_columns count];
+}
+
+- (NSGridRow *) rowAtIndex: (NSInteger)index
+{
+  return [_rows objectAtIndex: index];
+}
+
+- (NSInteger) indexOfRow: (NSGridRow *)row
+{
+  return [_rows indexOfObject: row];
+}
+
+- (NSGridColumn *) columnAtIndex: (NSInteger)index
+{
+  return [_columns objectAtIndex: index];
+}
+
+- (NSInteger) indexOfColumn: (NSGridColumn *)column
+{
+  return [_columns indexOfObject: column];
+}
+
+- (NSGridCell *) cellAtColumnIndex: (NSInteger)columnIndex rowIndex: (NSInteger)rowIndex
+{
+  NSGridColumn *col = [_columns objectAtIndex: columnIndex];
+  NSGridCell *c = [col cellAtIndex: rowIndex];
+  return c;
+}
+
+- (NSGridCell *) cellForView: (NSView*)view
+{
+  return nil;
+}
+
+- (NSGridRow *) addRowWithViews: (NSArray *)views
+{
+  return [self insertRowAtIndex: [_rows count]
+                      withViews: views];
+}
+
+- (NSGridRow *) insertRowAtIndex: (NSInteger)index withViews: (NSArray *)views
+{
+  NSGridRow *gr = [[NSGridRow alloc] init];
+
+  // Insert the row and release...
+  [_rows insertObject: gr atIndex: index];
+  RELEASE(gr);
+
+  // Insert views...
+  FOR_IN(NSView*, v, views)
+    {
+      NSGridCell *c = [[NSGridCell alloc] init];
+      [c setContentView: v];
+      RELEASE(v);
+      [gr _addCell: c];
+    }
+  END_FOR_IN(views);
+
+  // Refresh...
+  [self _refreshCells];
+  return gr;
+}
+
+- (void) moveRowAtIndex: (NSInteger)fromIndex toIndex: (NSInteger)toIndex
+{
+  [self _refreshCells];
+}
+
+- (void) removeRowAtIndex: (NSInteger)index
+{
+  [_rows removeObjectAtIndex: index];
+  [self _refreshCells];
```

Comment:
This also needs to change the columns by removing the corresponding cells.

Reply:
Yes, I am aware, working on that.

Reply:
Following your suggestion, so this is out of date.